Update language reference for non-trailing named arguments

garg000dhruv opened the pull request:
# Title

Update language reference for non-trailing named arguments

## Summary

Support has been added for non-trailing named arguments used in the correct order in C# 7.2. This PR updates the public C# language reference to explain the new contract.

Fixes #3240 

## Details

Since we're expanding use-cases for named arguments with C# 7.2, the public language reference will be update to explain the usages.

Moreover, a more sophisticated example replaces the current code snippet to cover all the variants that are now allowed along with ones that are still illegal. So instead of having a 2-parameter method, we'll have a 3-parameter method.

garg000dhruv committed:
Change example code explaining named arguments

Since we're expanding named arguments with C# 7.2, we need to have a
more sophisticated example to cover all the variants that are now
allowed along with ones that are still illegal. So instead of having a
2-parameter method, we'll have a 3-parameter method.

[Issue #: 3240]

dnfclas commented:
This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

garg000dhruv committed:
Change parameter name isShipped to hasShipped

garg000dhruv committed:
Update documentation for named arguments

Leveraging the updated code snippet, the named arguments section has
been updated with examples to explain non-trailing named arguments.

garg000dhruv commented:
Not sure if we want this dispersed in the documentation (or in a separate section of its own like "_What's new in C# 7.2_") but we can mention that the new usages in the examples are valid only for C# 7.2.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/named-and-optional-arguments.md:
```diff
@@ -50,28 +50,32 @@ translation.priority.ht:
  Named arguments also improve the readability of your code by identifying what each argument represents.  
   
- A named argument can follow positional arguments, as shown here.  
+ Named arguments are valid as long as they're used in the correct position, as shown here.
```
This is a bit misleading.  Named arguments are valid as long as 
1. They are not followed by any positional arguments.
-or-
2. They are used in the correct position.

The statement above implies they must always be in the correct order.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/named-and-optional-arguments.md:
```diff
@@ -50,28 +50,32 @@ translation.priority.ht:
+ `PrintOrderDetails(sellerName: "Foo", orderNum: 31, true);`
   
- `//CalculateBMI(weight: 123, 64);`  
+ However, out-of-order named arguments are invalid. The following statements cause a compiler error.
```
out-of-order named arguments are invalid **only** if they are followed by positional arguments.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/codesnippet/CSharp/named-and-optional-arguments_1.cs:
```diff
@@ -3,22 +3,27 @@ class NamedExample
         {
             // The method can be called in the normal way, by using positional arguments.
-            Console.WriteLine(CalculateBMI(123, 64));
+            PrintOrderDetails("Foo", 31, true);
```
In all locations where you call `PrintOrderDetails`, let's change the order name to a fake name that has some meaning: examples might be "gift order", "inventory replacement", or "recurring order".

garg000dhruv replied:
Makes sense. Updated the code snippet to use fake yet contextual entries.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/codesnippet/CSharp/named-and-optional-arguments_1.cs:
```diff
@@ -3,22 +3,27 @@ class NamedExample
         {
             // The method can be called in the normal way, by using positional arguments.
-            Console.WriteLine(CalculateBMI(123, 64));
+            PrintOrderDetails("Foo", 31, true);
```
One aspect of the previous sample that I liked was that the arguments had the same type. That showed that named arguments can be used to disambiguate arguments when calling a method.

A great example for this is throwing a `ArgumentException` or an `ArgumentNullException`. The `paramName` and `Message` arguments are easily reversed.

garg000dhruv replied:
Good point, I'll change up the example to contain this.

BillWagner commented:
@garg000dhruv I like dispersing the new features in throughout the topic. It reads better.

That said, I'd like to make sure that when you introduce the 7.2 features the first time, that you mention the version.
Something like "Starting with C# 7.2, you can...."

BillWagner reviewed (changes requested):
Thanks for submitting this @garg000dhruv Overall, this looks great.

I know this is tagged as WIP, but I gave it a good review and have a few suggestions. Let me know when you're ready for another look.

garg000dhruv committed:
Change code sample for named arguments

Based on PR feedback, the method in the code sample has been changed
to take in parameters that are the same type (string) to further amplify
the need of named arguments in the documentation.

garg000dhruv committed:
Update documentation based on PR feedback

garg000dhruv commented:
@BillWagner Addressed your feedback. Please take a look if things make sense.

BillWagner commented:
@garg000dhruv The changes so far make sense.  I like the direction.  Go ahead and finish up, polish this, and we'll be ready.

garg000dhruv commented:
@BillWagner I submitted this as the final polished version and removed _WIP_ from the title. So on my side, we're ready.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/named-and-optional-arguments.md:
```diff
@@ -50,28 +50,36 @@ translation.priority.ht:
+
+ `PrintOrderDetails("Gift Shop", 31, productName: "Red Mug");`
+
+- _starting with C# 7.2_, they're used in the correct position.
```
I would add a a note here that the 2nd argument is not named. Just to call it out explicitly and make it really clear what you are demonstrating.

jcouv replied:
It's not just that they are used in the correct position, but that they are not preceded by an argument that is used out-of-position.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/named-and-optional-arguments.md:
```diff
@@ -50,28 +50,36 @@ translation.priority.ht:
- `//CalculateBMI(weight: 123, 64);`  
+ However, out-of-order named arguments are invalid if they're followed by positional arguments. The following statements cause a compiler error.
+
+ `// PrintOrderDetails(productName: "Red Mug", 31, "Gift Shop");`
```
I'd remove the comments here, because it is inline code and won't be validated by our CI builds. (That way we get syntax highlighting.

I'd also recommend a comment with the generated error above the line.  We usually use:

```csharp
// This generates CSnnnn: CSnnnn error message.
BadCodeHere();
```

garg000dhruv replied:
Sorry for the delay. Been busy this week. Will work on this over the weekend and push out an update.

BillWagner replied:
Sounds good. Just @ - mention me when you finish and I'll give it a look.

BillWagner reviewed (changes requested):
Overall, this is great @garg000dhruv 

I've got a couple comments inline.

And, I'd really like to see an ArgumentException sample, as I mentioned in my first review.  Maybe something where the Seller is white space?

svick commented:
Does it make sense to document unreleased features? AFAIK, C# 7.2 hasn't been released yet, which means that:

1. Most people won't be able to use this for now.
2. There is still the possibility that what's in the final version of C# 7.2 will be different.

BillWagner commented:
@svick Good points.  

> 1. Most people won't be able to use this for now.

It is available in public preview. But yes, many people won't use it yet.

> 2. There is still the possibility that what's in the final version of C# 7.2 will be different.

Possibly. I met with @jcouv a couple weeks ago. We discussed the 'window' for when to publish new  content. We landed that we will use the window between when a feature branch is merged into master on the roslyn repo and the final release. This minimizes potential changes, and gives more developers the chance to read about and learn about the upcoming features.

That said, we should make sure these new features are all noted with the version when a feature first appears, and when it preview that should be noted as well.

jcouv commented:
@BillWagner Once 15.5 preview 2 is out, we will only fix severe compiler bugs. The new language features are extremely unlikely to change at that point (ie. we could publish the docs for C# 7.2).

garg000dhruv committed:
Point out C# 7.2 example with an explicit note

Per the review feedback, a note has been added when mentioning the C#
7.2 specific addition to named arguments so that it is explicitly clear
to the reader what the example entails.

garg000dhruv committed:
Specify the compiler error thrown for the bad example

garg000dhruv committed:
Add ArgumentException sample to code snippet

Per PR feedback, an ArgumentException was added to the example method
since both sellerName and productName are string types. This further
illustrates the point that named arguments enhance the readability of
the code.

garg000dhruv commented:
@BillWagner Addressed your feedback. 

The ```Argumentxception``` sample is in the code snippet along with some context in the documentation. We can add example method calls comparing the named and positional argument approaches to further illustrate the point but I thought the textual explanation should be enough. Let me know if you think differently.

garg000dhruv commented:
@BillWagner Not sure if you missed this. Just a ping to let you know that I've sent out the latest changes in this PR.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/codesnippet/CSharp/named-and-optional-arguments_1.cs:
```diff
@@ -3,22 +3,32 @@ class NamedExample
-            return (weight * 703) / (height * height);
+            if (string.IsNullOrWhiteSpace(sellerName))
+            {
+                throw new ArgumentException("Seller name cannot be null or empty.", nameof(sellerName));
```
Here, I would use the named arguments for clarity:

```csharp
throw new ArgumentException(message: "Seller name cannot be null or empty.", paramName: nameof(sellerName));
```

BillWagner reviewed (changes requested):
I like the updates.  One quick change, and then I'll merge it.

Thanks for helping us get this updated @garg000dhruv  we appreciate it.

BillWagner commented:
Thanks for the ping.  Crazy day today, so I just got to review it now.

Overall, it's great. One quick change and it's ready to :shipit:

garg000dhruv committed:
Add named arguments to exception in code snippet

garg000dhruv commented:
@BillWagner  Yay getting closer to my **first contribution** to the docs! Sent out the change requested.
